KAFKA-17674: Fix bug on update positions of newly added partitions

lianetm opened the pull request:
Fix to ensure that we keep the same set of initialization partitions within a single run of updateFetchPositions. This ensures that we don't reset positions using partition offsets for partitions that may be assigned while the updateFetchPositions is half-way (ie. after retrieving committed offsets but before the call to resetPositions). 

Before this PR, we could mistakenly reset positions to the partition offsets for a partition assigned that had committed offsets: fetch committed (not including the partition that wasn't assigned yet) + partition assigned + resetPositions(would include the partition that has been assigned and requires a position).

lianetm committed:
keep initial set of initializing partitions

lianetm committed:
adding test

lianetm committed:
remove extra line

lianetm commented:
Hello @chia7712, would you have a chance to take a look at this one maybe? Thanks!

chia7712 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java:
```diff
@@ -283,14 +283,15 @@ private CompletableFuture<Void> updatePositionsWithOffsets(long deadlineMs) {
+        final Set<TopicPartition> initializingPartitions = subscriptionState.initializingPartitions();
         if (commitRequestManager != null) {
-            CompletableFuture<Void> refreshWithCommittedOffsets = initWithCommittedOffsetsIfNeeded(deadlineMs);
+            CompletableFuture<Void> refreshWithCommittedOffsets = initWithCommittedOffsetsIfNeeded(initializingPartitions, deadlineMs);
```
Perhaps `initWithCommittedOffsetsIfNeeded` could return a `CompletableFuture<Set<TopicPartition>>` to include the initializing partitions, as the `refreshOffsets` method can provide this information.

lianetm replied:
I see, but doesn't it feel more complicated to rely on a return value here vs an explicit input parameter?

Also, if we get the info from `refreshOffsets` we would need to get it based on the response (if it has response for tp1 and tp2 we can safely know that those are the ones we requested). Is this what you had in mind? 

Based on an explicit input param, (what we have now), we get the `subscription.initializingPartitions()` once at the beginning of the `updatePositionsWithOffsets` and pass it as input param to the 2 phases that need it:
1. initWithCommittedOffsetsIfNeeded(initializingPartitions, ...)
2. initWithPartitionOffsetsIfNeeded(initializingPartitions)

Am I maybe missing an issue with this approach of the explicit input param or is it that you find the return value clearer?

chia7712 replied:
> Am I maybe missing an issue with this approach of the explicit input param or is it that you find the return value clearer?

I was considering that initWithCommittedOffsetsIfNeeded might adjust the initializing partitions, similar to KAFKA-16966, but that isn't happening at the moment. I'm +1 on the current approach, as it’s simpler and more readable.

chia7712 reviewed (commented):
> we don't reset positions using partition offsets for partitions that may be assigned while the updateFetchPositions is half-way (ie. after retrieving committed offsets but before the call to resetPositions).

Sorry, could you explain how to reproduce the issue? I guess Consumer users are experiencing timeouts, but the CheckAndUpdatePositionsEvent is still processing. After that, users manually assigning positions to the partitions, right?

lianetm commented:
> Sorry, could you explain how to reproduce the issue? I guess Consumer users are experiencing timeouts, but the CheckAndUpdatePositionsEvent is still processing. After that, users manually assigning positions to the partitions, right?

Exactly, basically we can repro this by getting a new assigned partition while the CheckAndUpdatePositionsEvent is being processed (becoming an issue if we get the newly assigned partition after the OffsetFetch was sent, but before resetPositions). So given that initializingPartitions can change anytime with a new assignment, the goal is to ensure that we keep the same one while processing a single CheckAndUpdatePositionsEvent (whatever partition is added will get a position on the next CheckAndUpdatePositionsEvent).

For context, this came out on a new system test (not yet in AK) that would sometimes fail with this. Here is the sequence in the logs that shows the problematic situation:

```
[2024-10-01 21:04:58,244] INFO [Consumer instanceId=test_group_id-instance-1, clientId=consumer-test_group_id-test_group_id-instance-1, groupId=test_group_id] Reconciling assignment with local epoch EXhOefVLQ7K5SMB3tMyuDg
	Member:                                    0
	Assigned partitions:                       [test_topic-0]
	Current owned partitions:                  []
	Added partitions (assigned - owned):       [test_topic-0]
	Revoked partitions (owned - assigned):     []
 (org.apache.kafka.clients.consumer.internals.MembershipManagerImpl)
…
[2024-10-01 21:05:03,251] DEBUG [Consumer instanceId=test_group_id-instance-1, clientId=consumer-test_group_id-test_group_id-instance-1, groupId=test_group_id] Refreshing committed offsets for partitions [test_topic-0] (org.apache.kafka.clients.consumer.internals.AsyncKafkaConsumer)
[2024-10-01 21:05:03,251] TRACE For telemetry state SUBSCRIPTION_NEEDED, returning the value 294947 ms;  (org.apache.kafka.common.telemetry.internals.ClientTelemetryReporter)
[2024-10-01 21:05:03,252] DEBUG [Consumer instanceId=test_group_id-instance-1, clientId=consumer-test_group_id-test_group_id-instance-1, groupId=test_group_id] Enqueuing offset fetch request for partitions: [test_topic-0] (org.apache.kafka.clients.consumer.internals.CommitRequestManager)
…
[2024-10-01 21:05:03,256] DEBUG [Consumer instanceId=test_group_id-instance-1, clientId=consumer-test_group_id-test_group_id-instance-1, groupId=test_group_id] Sending OFFSET_FETCH request with header RequestHeader(apiKey=OFFSET_FETCH, apiVersion=9, clientId=consumer-test_group_id-test_group_id-instance-1, correlationId=15, headerVersion=2) and timeout 30000 to node 2147483646: OffsetFetchRequestData(groupId='', topics=[], groups=[OffsetFetchRequestGroup(groupId='test_group_id', memberId='EXhOefVLQ7K5SMB3tMyuDg', memberEpoch=2, topics=[OffsetFetchRequestTopics(name='test_topic', partitionIndexes=[0])])], requireStable=true) (org.apache.kafka.clients.NetworkClient)
…
[2024-10-01 21:05:03,257] INFO [Consumer instanceId=test_group_id-instance-1, clientId=consumer-test_group_id-test_group_id-instance-1, groupId=test_group_id] Reconciling assignment with local epoch EXhOefVLQ7K5SMB3tMyuDg
	Member:                                    1
	Assigned partitions:                       [test_topic-0, test_topic-1]
	Current owned partitions:                  [test_topic-0]
	Added partitions (assigned - owned):       [test_topic-1]
	Revoked partitions (owned - assigned):     []
 (org.apache.kafka.clients.consumer.internals.MembershipManagerImpl)
…
[2024-10-01 21:05:03,264] DEBUG [Consumer instanceId=test_group_id-instance-1, clientId=consumer-test_group_id-test_group_id-instance-1, groupId=test_group_id] Received OFFSET_FETCH response from node 2147483646 for request with header RequestHeader(apiKey=OFFSET_FETCH, apiVersion=9, clientId=consumer-test_group_id-test_group_id-instance-1, correlationId=15, headerVersion=2): OffsetFetchResponseData(throttleTimeMs=0, topics=[], errorCode=0, groups=[OffsetFetchResponseGroup(groupId='test_group_id', topics=[OffsetFetchResponseTopics(name='test_topic', partitions=[OffsetFetchResponsePartitions(partitionIndex=0, committedOffset=324, committedLeaderEpoch=-1, metadata='', errorCode=0)])], errorCode=0)]) (org.apache.kafka.clients.NetworkClient)
…
[2024-10-01 21:05:03,268] DEBUG [Consumer instanceId=test_group_id-instance-1, clientId=consumer-test_group_id-test_group_id-instance-1, groupId=test_group_id] Creating ListOffset request ListOffsetsRequestData(replicaId=-1, isolationLevel=0, topics=[ListOffsetsTopic(name='test_topic', partitions=[ListOffsetsPartition(partitionIndex=1, currentLeaderEpoch=0, timestamp=-2, maxNumOffsets=1)])]) for broker ducker03:9092 (id: 1 rack: null) to reset positions (org.apache.kafka.clients.consumer.internals.OffsetsRequestManager) _==> This is what's wrong, it's requesting partitions offsets for partition 1 to reset positions for it, just because it's initializing after the the Offset_Fetch completed, but that's not because there are no offsets for it, it's because it was added after the request was sent._ 
```

lianetm commented:
Hey @chia7712 , thanks a lot for taking the time to check this one! Comments addressed.

chia7712 reviewed (approved):
LGTM
